spi/data-plane/data-plane-spi/src/main/java/org/eclipse/edc/connector/dataplane/spi/DataFlow.java

```diff
@@ -103,8 +107,12 @@ public void transitToNotified() {
         transitionTo(NOTIFIED.code());
     }
 
-    public void transitToTerminated() {
+    public void transitToTerminated(@Nullable String reason) {
         transitionTo(TERMINATED.code());
+        if (reason != null) {
+            properties = new HashMap<>(properties); //properties could be immutable
```

Comment:
this shoulnd't be the case, I would change the instantiation at line 47 and the way the `properties` builder behaves, otherwise we'll need to instantiate a new map every time we need to add a property (in this object this is actually the only place where we do that, but generally speaking an entity should be mutable by default.)